File: bridge/core/dom/child_node_list.cc
/*
 * Copyright (C) 2019-2022 The Kraken authors. All rights reserved.
 * Copyright (C) 2022-present The WebF authors. All rights reserved.
 */

#include "child_node_list.h"
#include "bindings/qjs/cppgc/gc_visitor.h"

namespace webf {

ChildNodeList::ChildNodeList(ContainerNode* parent) : parent_(parent), NodeList(parent->ctx()) {}
ChildNodeList::~ChildNodeList() = default;

Node* ChildNodeList::VirtualOwnerNode() const {
  return &OwnerNode();
}

Node* ChildNodeList::item(unsigned index, ExceptionState& exception_state) const {
  return collection_index_cache_.NodeAt(*this, index);
}

bool ChildNodeList::NamedPropertyQuery(const AtomicString& key, ExceptionState& exception_state) {
  int32_t index = std::stoi(key.ToStdString(ctx()));
  return collection_index_cache_.NodeAt(*this, index);
}

void ChildNodeList::NamedPropertyEnumerator(std::vector<AtomicString>& names, ExceptionState& exception_state) {
  uint32_t size = collection_index_cache_.NodeCount(*this);
  for (int i = 0; i < size; i++) {
    names.emplace_back(AtomicString(ctx(), std::to_string(i)));
  }
}

Node* ChildNodeList::TraverseForwardToOffset(unsigned offset, Node& current_node, unsigned& current_offset) const {
  assert(current_offset < offset);
  assert(OwnerNode().childNodes() == this);
  assert(&OwnerNode() == current_node.parentNode());
  for (Node* next = current_node.nextSibling(); next; next = next->nextSibling()) {
    if (++current_offset == offset)
      return next;
  }
  return nullptr;
}

Node* ChildNodeList::TraverseBackwardToOffset(unsigned offset, Node& current_node, unsigned& current_offset) const {
  assert(current_offset > offset);
  assert(OwnerNode().childNodes() == this);
  assert(&OwnerNode() == current_node.parentNode());
  for (Node* previous = current_node.previousSibling(); previous; previous = previous->previousSibling()) {
    if (--current_offset == offset)
      return previous;
  }
  return nullptr;
}

void ChildNodeList::Trace(GCVisitor* visitor) const {
  visitor->Trace(parent_);
  collection_index_cache_.Trace(visitor);
  NodeList::Trace(visitor);
}

}  // namespace webf

File: webf/lib/src/dom/child_node_list.dart
/*
 * Copyright (C) 2022-present The WebF authors. All rights reserved.
 */

import 'package:flutter/widgets.dart' as flutter;
import 'node.dart';
import 'element.dart';
import 'node_list.dart';
import 'container_node.dart';
import 'collection_index_cache.dart';

class _ChildNodeListIterator extends Iterator<Node> {
  final ChildNodeList collection;
  int _index = -1;
  Node? _current;

  _ChildNodeListIterator(this.collection);

  @override
  Node get current {
    return _current!;
  }

  @override
  bool moveNext() {
    if (_current == null) {
      _current = collection.ownerNode.firstChild;
      if (_current == null) return false;
    } else {
      _current = collection.traverseForwardToOffset(_index + 1, _current!, [_index]);
    }
    _index++;
    return _current != null;
  }
}

class ChildNodeList extends NodeList {
  ChildNodeList(ContainerNode rootNode)
      : _owner = rootNode,
        _collectionIndexCache = CollectionIndexCache<ChildNodeList, Node>();

  final ContainerNode _owner;
  final CollectionIndexCache<ChildNodeList, Node> _collectionIndexCache;
  List<flutter.Widget>? _cachedWidgetList;

  @override
  int get length => _collectionIndexCache.nodeCount(this);

  @override
  bool get isChildNodeList => true;

  @override
  Node get ownerNode => _owner;

  Node? item(int index) {
    return _collectionIndexCache.nodeAt(this, index);
  }

  List<flutter.Widget> toWidgetList() {
    // if (_cachedWidgetList != null) {
    //   return _cachedWidgetList!;
    // }
    List<flutter.Widget> result = map((node) => node.toWidget()).toList();
    return result;
  }

  void childrenChanged(ChildrenChange change) {
    if (change.isChildInsertion()) {
      _collectionIndexCache.nodeInserted();
    } else if (change.isChildRemoval()) {
      _collectionIndexCache.nodeRemoved();
    } else {
      _collectionIndexCache.invalidate();
    }
    _cachedWidgetList = null;
  }

  void invalidateCache() {
    _collectionIndexCache.invalidate();
    _cachedWidgetList = null;
  }

  bool get canTraverseBackward => true;

  T? traverseToFirst<T extends Node>() => _owner.firstChild as T?;
  T? traverseToLast<T extends Node>() => _owner.lastChild as T?;

  T? traverseForwardToOffset<T extends Node>(int offset, Node currentNode, List<int> currentOffset) {
    assert(currentOffset.length == 1);
    assert(currentOffset[0] < offset);
    assert(ownerNode.childNodes == this);
    assert(ownerNode == currentNode.parentNode);
    Node? next = currentNode.nextSibling;
    while (next != null) {
      currentOffset[0]++;
      if (currentOffset[0] == offset) {
        return next as T?;
      }
      next = next.nextSibling;
    }
    return null;
  }

  T? traverseBackwardToOffset<T extends Node>(int offset, Node currentNode, List<int> currentOffset) {
    assert(currentOffset.length == 1);
    assert(currentOffset[0] > offset);
    assert(ownerNode.childNodes == this);
    assert(ownerNode == currentNode.parentNode);
    Node? previous = currentNode.previousSibling;
    while (previous != null) {
      currentOffset[0]--;
      if (currentOffset[0] == offset) {
        return previous as T?;
      }
      previous = previous.previousSibling;
    }
    return null;
  }

  @override
  bool get isEmpty => _collectionIndexCache.isEmpty(this);

  @override
  bool get isNotEmpty => !_collectionIndexCache.isEmpty(this);

  @override
  Iterator<Node> get iterator => _ChildNodeListIterator(this);

  @override
  String hashKey() {
    String key = '';
    for (int i = 0; i < length; i ++) {
      key += '_' + elementAt(i).hashKey;
    }
    return key;
  }
}
